regproxy/proxy.go

```diff
@@ -1,115 +1,143 @@
-			if len(parts) < 2 {
-				err = errors.New("invalid status line")
-				return
-			}
-			status, err = strconv.Atoi(parts[1])
+// Serve waits for TCP connections and pipes data received from the connection
+// to BK via the gRPC server.
+func (r *RegistryProxy) Serve(ctx context.Context) {
+	wg := sync.WaitGroup{}
+	defer func() {
+		wg.Wait()
+		close(r.errCh)
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			r.errCh <- ctx.Err()
+			return
+		default:
+			conn, err := r.ln.Accept()
 			if err != nil {
+				if !r.done.Load() {
+					r.errCh <- errors.Wrap(err, "failed to accept")
+				}
 				return
 			}
-			continue
-		}
-		parts := strings.Split(line, ": ")
-		if len(parts) < 2 {
-			err = errors.New("invalid header format")
-			return
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				r.errCh <- r.handle(ctx, conn)
+			}()
 		}
-		header.Add(parts[0], parts[1])
 	}
+}
 
-	err = sc.Err()
+func (r *RegistryProxy) Close() {
+	r.done.Store(true)
+	r.ln.Close()
+}
 
-	return
+func (r *RegistryProxy) Err() <-chan error {
+	return r.errCh
 }
 
-// serve the HTTP request by writing it to BK via a streaming gRPC request. The
-// request will be reconstituted on the other end and forwarded to the embedded
-// registry and back again.
-func (r *RegistryProxy) serve(w http.ResponseWriter, req *http.Request) error {
+func (r *RegistryProxy) handle(ctx context.Context, conn net.Conn) error {
+	defer conn.Close()
 
-	stream, err := r.cl.Proxy(req.Context())
+	stream, err := r.cl.Proxy(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to send proxy request: %w", err)
+		return errors.Wrap(err, "failed to create proxy stream")
 	}
 
 	rw := registry.NewStreamRW(stream)
+	eg, _ := errgroup.WithContext(ctx)
 
-	err = req.WriteProxy(rw)
-	if err != nil {
-		return fmt.Errorf("failed to write request: %w", err)
-	}
+	eg.Go(func() error {
+		_, err = copyWithDeadline(conn, rw)
+		if err != nil {
+			return errors.Wrap(err, "failed to write to stream")
+		}
+		err = stream.CloseSend()
+		if err != nil {
+			return errors.Wrap(err, "failed to close stream")
+		}
+		return nil
+	})
 
-	err = stream.CloseSend()
-	if err != nil {
-		return fmt.Errorf("failed to send close: %w", err)
-	}
+	eg.Go(func() error {
+		_, err = io.Copy(conn, rw)
+		if err != nil {
+			return errors.Wrap(err, "failed to read from stream")
+		}
+		return nil
+	})
 
-	status, header, err := parseHeader(rw)
+	err = eg.Wait()
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to wait")
 	}
 
-	for key, vals := range header {
-		for _, val := range vals {
-			w.Header().Add(key, val)
+	return nil
+}
+
+func copyWithDeadline(conn net.Conn, w io.Writer) (int64, error) {
+	var t int64
+	for {
+		err := conn.SetReadDeadline(time.Now().Add(readDeadline))
+		if err != nil {
+			return 0, err
 		}
+		buf := make([]byte, 1024)
+		n, err := conn.Read(buf)
+		if err != nil {
+			if errors.Is(err, io.EOF) || isNetTimeout(err) {
+				break
+			}
+			return 0, err
```

Comment:
nit: IO commands tend to return the number of successfully read/written bytes along with the error, to allow the caller to be more graceful in known how many bytes were processed before hand.

I'm aware the code that calls this function doesn't use it, but it future-proofs it in case we decide it's needed down the road.

```suggestion
			return t, err
```